### PR TITLE
[1.27]  kubelet: devices: skip allocation for running pods #118635 

### LINKED ISSUE #118559: Running pods with devices are terminated if kubelet is restarted
### What happened?

In KubeVirt project, we now see a regression when running on Kubernetes `1.25.10 | 1.26.5 | 1.27.2`. If kubelet is restarted on a node, then all the existing and running workloads that use devices are terminated with `UnexpectedAdmissionError`:
```
Warning  UnexpectedAdmissionError  45s   kubelet            Allocate failed due to no healthy devices present; cannot allocate unhealthy devices devices.kubevirt.io/kvm, which is unexpected
Normal   Killing                   42s   kubelet            Stopping container compute
```
KubeVirt runs virtual machines inside pods and uses a device plugin to advertise e.g. `/dev/kvm` on the nodes.

Presumably, this PR changed the behavior: https://github.com/kubernetes/kubernetes/pull/116376
Original issue: https://github.com/kubernetes/kubernetes/issues/109595

### What did you expect to happen?

A potential restart of kubelet should not interrupt the running workloads.

### How can we reproduce it (as minimally and precisely as possible)?
with KubeVirt:

- run a KubeVirt VM
- `pkill kubelet`
- observe that the workload pod gets terminated

or with https://github.com/k8stopologyawareschedwg/sample-device-plugin

- make deploy
- make test-both
- pkill kubelet
- the pod gets restarted

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

This affects the 1.25.x, 1.26.x and 1.27.x branches.

1.25.10 | 1.26.5 | 1.27.2

</details>

### Cloud provider

<details>

N/A

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:
Cherry-pick of #118635 to branch `release-1.27`. Cherry pick per se done using `hack/cherry_pick_pull.sh`

#### Original description
When kubelet initializes, runs admission for pods and possibly allocate requested resources. We need to distinguish between node reboot (no containers running) versus kubelet restart (containers potentially running).

Running pods should always survive kubelet restart. This means that device allocation on admission should not be attempted, because if a container requires devices and is still running when kubelet is restarting, that container already has devices allocated and working.

Thus, we need to properly detect this scenario in the allocation step and handle it explicitely. We need to inform the devicemanager about which pods are already running.

#### Which issue(s) this PR fixes:
Fixes #118559

#### Special notes for your reviewer:
Implements the first approach [proposed](https://github.com/kubernetes/kubernetes/issues/118559#issuecomment-1584636260) in [the thread](https://github.com/kubernetes/kubernetes/issues/118559#issuecomment-1584802420), so we make the devicemanager treat running pod differently.

This approach was chosen because it seems simpler to make self-contained and easier to backport.

The devicemanager already tracks (with the help of the checkpoint files) which containers got devices assigned to them, which by definition means these containers passed its admission. The missing bit is safely learning which container are already running when initializing, and for that we extend the [existing `buildContainerMapFromRuntime`](https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.2/pkg/kubelet/cm/container_manager_linux.go#L567)

#### Does this PR introduce a user-facing change?
```release-note
Fixes regression in 1.27.2 causing running pods with devices to be terminated if kubelet is restarted
```

### CODE REVIEW

**Comment by ffromani** [comment]
/test pull-kubernetes-unit-go-compatibility

unrelated failure: `k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset: TestTooWide expand_less`

**Comment by ffromani** [comment]
/test pull-kubernetes-e2e-gce-device-plugin-gpu

setup failure?

**Comment by ffromani** [comment]
/test pull-kubernetes-e2e-capz-windows-1-27

unrelated failure: `/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/conformance_test.go`

**Comment by ffromani** [comment]
/retest-required

apiserver is not affected by this change

**Comment by ffromani** [comment]
I'm confident the failures are unrelated and/or flakes and I'm gonna file issues for them ASAP

**Comment by bobbypage** [comment]
What is the reason for the pull-kubernetes-e2e-gce-device-plugin-gpu failure? I see the job is passing in master - https://testgrid.k8s.io/sig-release-master-blocking#gce-device-plugin-gpu-master

**Comment by ffromani** [comment]
> What is the reason for the pull-kubernetes-e2e-gce-device-plugin-gpu failure? I see the job is passing in master - https://testgrid.k8s.io/sig-release-master-blocking#gce-device-plugin-gpu-master

fair point. Investigating

**Comment by ffromani** [comment]
crosschecking this PR (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119432/pull-kubernetes-e2e-gce-device-plugin-gpu/1683770620283719680/) with a dummy one (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119590/pull-kubernetes-e2e-gce-device-plugin-gpu/1684154811563380736/) points in the direction of the lane configuration. I'll check test-infra

**Comment by ffromani** [comment]
> What is the reason for the pull-kubernetes-e2e-gce-device-plugin-gpu failure? I see the job is passing in master - https://testgrid.k8s.io/sig-release-master-blocking#gce-device-plugin-gpu-master

I don't know yet :) I'm still investigating. After a first pass, the job definition doesn't look obviously wrong. But I'm very rusty in this area so I may be miss something. I'll keep looking.

Another fact emerged during the investigation is that, in addition to the job being broken which should be obviously addressed, is that the `GPUDevicePlugin` tests don't really exercise the flow we care in this PR. There are no GPUDevicePlugin tests in `test/e2e_node` at all:
```
1023 12:33:22 fromani@musashi3 ~/go/src/k8s.io/kubernetes $ git log --oneline -1
6d31f4b31ba (HEAD -> release-1.27, upstream/release-1.27) Merge pull request #119519 from jingxu97/automated-cherry-pick-of-#118451-upstream-release-1.27
1024 12:33:26 fromani@musashi3 ~/go/src/k8s.io/kubernetes $ git grep GPUDevicePlugin
test/e2e/framework/gpu/gpu_util.go:24:  // GPUDevicePluginDSYAML is the official Google Device Plugin Daemonset NVIDIA GPU manifest for GKE
test/e2e/framework/gpu/gpu_util.go:27:  GPUDevicePluginDSYAML = "https://raw.githubusercontent.com/kubernetes/kubernetes/master/cluster/addons/device-plugins/nvidia-gpu/daemonset.yaml"
test/e2e/scheduling/nvidia-gpus.go:224:var _ = SIGDescribe("[Feature:GPUDevicePlugin]", func() {
test/e2e/scheduling/nvidia-gpus.go:323:var _ = SIGDescribe("GPUDevicePluginAcrossRecreate [Feature:Recreate]", func() {
test/e2e/windows/device_plugin.go:42:var _ = SIGDescribe("[Feature:GPUDevicePlugin] Device Plugin", func() {
test/e2e_node/image_list.go:85: if gpuDevicePluginImage, err := getGPUDevicePluginImage(ctx); err != nil {
test/e2e_node/image_list.go:215:// getGPUDevicePluginImage returns the image of GPU device plugin.
test/e2e_node/image_list.go:216:func getGPUDevicePluginImage(ctx context.Context) (string, error) {
test/e2e_node/image_list.go:217:        ds, err := e2emanifest.DaemonSetFromURL(ctx, e2egpu.GPUDevicePluginDSYAML)
```

I'm inclined to do the debugging of the jobs in parallel and not block this PR because of that, please reviewers share your thoughts

**Comment by ffromani** [comment]
filed: https://github.com/kubernetes/kubernetes/issues/119730

**Comment by ffromani** [comment]
/test pull-kubernetes-e2e-capz-windows-1-27

the 1.26 lane seems to be fixed

**Comment by swatisehgal** [comment]
/test pull-kubernetes-e2e-gce-device-plugin-gpu
I see this job passed for https://github.com/kubernetes/kubernetes/pull/119590 so lets try again!

/test pull-kubernetes-e2e-capz-windows-1-27
this one has been failing consistently but failures are unrelated to this PR. Let's see if another attempt helps.

**Comment by ffromani** [comment]
from what I gathered, the `pull-kubernetes-e2e-gce-device-plugin-gpu` fails at setup stage because the installation process fail to setup the devices correctly, it seems at driver compilation stage?
I'm not knowledgeable enough in this flow to debug further

**Comment by ffromani** [comment]
rebased to fix a conflict

**Comment by ffromani** [comment]
/retest

unrelated failures

**Comment by ffromani** [comment]
/test pull-kubernetes-e2e-capz-windows-1-27

authentication error hardly caused by changes in device manager

**Comment by ffromani** [comment]
/test pull-kubernetes-e2e-capz-windows-1-27

according to 1.26 result, it seems a flake (https://github.com/kubernetes/kubernetes/issues/119734)

**Comment by ffromani** [comment]
/test pull-kubernetes-e2e-capz-windows-1-27

ErrImagePull seems really unrelated to this PR

**Comment by ffromani** [comment]
it seems we need at least https://github.com/kubernetes/test-infra/pull/30348

**Comment by jsturtevant** [comment]
@ffromani sorry for the issues with the windows job. Thanks for reporting.  It is running and passing now

**Comment by ffromani** [comment]
> @ffromani sorry for the issues with the windows job. Thanks for reporting. It is running and passing now

no problem at all! it took me a while to get a clue about what was broken, but appy it was resolved and we have better windows jobs

**Review by swatisehgal — COMMENTED** [review]
/lgtm

Re-applying as the label was removed due to rebase.

**Comment by bobbypage** [comment]
regarding https://github.com/kubernetes/kubernetes/pull/119432#issuecomment-1655462848

> I'm inclined to do the debugging of the jobs in parallel and not block this PR because of that, please reviewers share your thoughts

+1, it looks like the test is failing on presubmits on your empty PR (https://github.com/kubernetes/kubernetes/pull/119590), so I agree we should not block this PR on that. I believe I found the issue there with the config (see https://github.com/kubernetes/test-infra/pull/30450), which we can address separately.

**Comment by ffromani** [comment]
> regarding [#119432 (comment)](https://github.com/kubernetes/kubernetes/pull/119432#issuecomment-1655462848)
> 
> > I'm inclined to do the debugging of the jobs in parallel and not block this PR because of that, please reviewers share your thoughts
> 
> +1, it looks like the test is failing on presubmits on your empty PR (#119590), so I agree we should not block this PR on that. I believe I found the issue there with the config (see [kubernetes/test-infra#30450](https://github.com/kubernetes/test-infra/pull/30450)), which we can address separately.

Thanks @bobbypage ! it seems your fix was merged and worked like a charm, awesome!
Now I'll review myself that change carefully because I was not aware of the interdependencies and of the settings you used. TIL!

**Comment by retr0h** [comment]
@ffromani any idea when this will make it into 1.27? 

We have been hitting this for a while now, and encounter `UnexpectedAdmissionError` on pods which use devices made available by the generic device plugin when the kubelet is restarted.

**Comment by ffromani** [comment]
> @ffromani any idea when this will make it into 1.27?
> 
> We have been hitting this for a while now, and encounter `UnexpectedAdmissionError` on pods which use devices made available by the generic device plugin when the kubelet is restarted.

Hi, please note that we (or, at very least, I :) ) want to keep the existing behavior added in https://github.com/kubernetes/kubernetes/issues/109595 , whose partial fix lead to the bug I'm fixing here. In other words, some pods WILL still be hitting UnexpectedAdmissionError in some cases.
 That said, I'm aiming for the september patch day. All blockers have been removed, but the backports have not been approved either.

**Review by mrunalp — APPROVED** [review]

**Comment by bobbypage** [comment]
/cc @kubernetes/release-managers 

Release managers, can you please take a look at this cherrypick? Thank you!

**Review by xmudrii — APPROVED** [review]
For RelEng:
/lgtm
/approve

**Comment by ffromani** [comment]
/test pull-kubernetes-unit

unrelated failures
image

**Comment by ffromani** [comment]
/test pull-kubernetes-unit-go-compatibility

unrelated failure

**Pull request merged**
